cli/src/commands/rebase.rs

```diff
@@ -873,17 +874,33 @@ fn move_commits(
             if let Some(new_child_parents) = new_children_parents.get(commit_id) {
                 new_child_parents.clone()
             }
-            // Commits in the target set should persist only rebased parents from the target
-            // sets.
-            else if let Some(target_commit_parents) =
-                target_commits_internal_parents.get(commit_id)
-            {
-                // If the commit does not have any parents in the target set, it is one of the
-                // commits in the root set, and should be rebased onto the new destination.
-                if target_commit_parents.is_empty() {
+            // Commit is in the target set.
+            else if target_commit_ids.contains(commit_id) {
+                // If the commit is a root of the target set, it should be rebased onto the new destination.
+                if target_roots.contains(commit_id) {
                     new_parent_ids.clone()
-                } else {
-                    target_commit_parents.clone()
+                }
+                // Otherwise:
+                // 1. Keep parents which are within the target set.
+                // 2. Replace parents which are outside the target set but are part of the
+                //    connected target set with their ancestor commits which are in the target
+                //    set.
+                // 3. Keep other parents outside the target set if they are not descendants of the
+                //    new children of the target set.
+                else {
+                    let mut new_parents = vec![];
+                    for parent_id in commit.parent_ids() {
+                        if target_commit_ids.contains(parent_id) {
+                            new_parents.push(parent_id.clone());
+                        } else if let Some(parents) =
+                                connected_target_commits_internal_parents.get(parent_id) {
+                            new_parents.extend(parents.iter().cloned());
+                        } else if !new_children.iter().any(|new_child| {
```

Comment:
I think this should be an error, but I'm not sure if we handle other loop cases differently.

Reply:
We have this code for handling loops already: https://github.com/martinvonz/jj/blob/19563fee7483ecd6b5f56200520ac81c84ea16c5/cli/src/commands/rebase.rs#L713-L743

I'm not sure it's worth having that code because it also seems like strange behavior to support but I don't feel strongly. My impression was that @bnjmnt4n also didn't feel strongly either way.

Reply:
> I think this should be an error, but I'm not sure if we handle other loop cases differently.

In this case, I feel like the user's intent in calling rebase `--before`/`--after` is to move the commits to the new location, making whatever changes to the commit graph are necessary, so this was acceptable to me. `--before`/`--after` allows you to move the root of the target set before its original parents, so I thought moving the other commits of the target set before any of its original parents is acceptable.